In [1]:
%matplotlib inline
%pip install -q transformers

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
#import seaborn as sns
from collections import Counter
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from transformers import pipeline, AutoModelForSequenceClassification, AutoTokenizer
from transformers import PreTrainedTokenizerFast
from tokenizers import BertWordPieceTokenizer

None of PyTorch, TensorFlow >= 2.0, or Flax have been found. Models won't be available and only tokenizers, configuration and file/data utilities can be used.


# Working with Text Lab
## Information retrieval, preprocessing, and feature extraction

In this lab, you'll be looking at and exploring European restaurant reviews. The dataset is rather tiny, but that's just because it has to run on any machine. In real life, just like with images, texts can be several terabytes long.

The dataset is located [here](https://www.kaggle.com/datasets/gorororororo23/european-restaurant-reviews) and as always, it's been provided to you in the `data/` folder.

### Problem 1. Read the dataset (1 point)
Read the dataset, get acquainted with it. Ensure the data is valid before you proceed.

How many observations are there? Which country is the most represented? What time range does the dataset represent?

Is the sample balanced in terms of restaurants, i.e., do you have an equal number of reviews for each one? Most importantly, is the dataset balanced in terms of **sentiment**?

In [3]:
european_restaurants = pd.read_csv("data/European Restaurant Reviews.csv")

In [4]:
european_restaurants.describe()

Country            Restaurant Name Sentiment Review Title Review Date  \
count     1502                       1502      1502         1502        1502   
unique       7                          7         2         1343         143   
top     France  The Frog at Bercy Village  Positive    Excellent  May 2014 •   
freq       512                        512      1237            9         108   

                                                   Review  
count                                                1502  
unique                                               1426  
top     I'm afraid we were very let down by this resta...  
freq                                                    4

In [5]:
european_restaurants.shape

(1502, 6)

In [13]:
european_restaurants.columns = european_restaurants.columns.str.lower().str.replace(" ","_")

In [14]:
european_restaurants.country.value_counts().unique

<bound method Series.unique of country
France     512
Italy      318
Morroco    210
Cuba       146
Poland     135
Russia     100
India       81
Name: count, dtype: int64>

In [17]:
european_restaurants.review_date.min(), european_restaurants.review_date.max()

('Apr 2012 •', 'Sept 2023 •')

In [18]:
european_restaurants.dtypes

country            object
restaurant_name    object
sentiment          object
review_title       object
review_date        object
review             object
dtype: object

In [23]:
european_restaurants.drop_duplicates()

country            restaurant_name sentiment  \
0     France  The Frog at Bercy Village  Negative   
1     France  The Frog at Bercy Village  Negative   
2     France  The Frog at Bercy Village  Negative   
3     France  The Frog at Bercy Village  Negative   
4     France  The Frog at Bercy Village  Negative   
...      ...                        ...       ...   
1492    Cuba   Old Square (Plaza Vieja)  Negative   
1493    Cuba   Old Square (Plaza Vieja)  Negative   
1494    Cuba   Old Square (Plaza Vieja)  Negative   
1495    Cuba   Old Square (Plaza Vieja)  Negative   
1496    Cuba   Old Square (Plaza Vieja)  Negative   

                                      review_title review_date  \
0                                     Rude manager  May 2024 •   
1                             A big disappointment  Feb 2024 •   
2                     Pretty Place with Bland Food  Nov 2023 •   
3         Great service and wine but inedible food  Mar 2023 •   
4        Avoid- Worst meal in Rome - possibly ever  Nov 2022 •   
...                                            ...         ...   
1492              It's nothing exciting over there  Oct 2016 •   
1493                                  Tourist Trap  Oct 2016 •   
1494  Over-renovated, or in a state of transition?  Oct 2016 •   
1495       The square is beautiful, but that's it.  Oct 2016 •   
1496               Great place to watch Cuba go by  Oct 2016 •   

                                                 review  
0     The manager became agressive when I said the c...  
1     I ordered a beef fillet ask to be done medium,...  
2     This is an attractive venue with welcoming, al...  
3     Sadly I  used the high TripAdvisor rating too ...  
4     From the start this meal was bad- especially g...  
...                                                 ...  
1492  We visited the Havana's Club Museum which is l...  
1493  Food and service was awful. Very pretty stop. ...  
1494  I was disappointed with the Plaza Vieja, to wh...  
1495  Beautiful in the square and up to two blocks o...  
1496  We were lucky enough to have an overnight stay...  

[1453 rows x 6 columns]

In [28]:
european_restaurants.review_date = european_restaurants.review_date.str[:-1]

#### Observations were 1502 and after drop duplicates they becaome 1453. Most represented country is France. Time range is form April 2012 till September 2023

### Problem 2. Getting acquainted with reviews (1 point)
Are positive comments typically shorter or longer? Try to define a good, robust metric for "length" of a text; it's not necessary just the character count. Can you explain your findings?

In [33]:
european_restaurants.groupby("sentiment").review.value_counts()

sentiment  review                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       

### Problem 3. Preprocess the review content (2 points)
You'll likely need to do this while working on the problems below, but try to synthesize (and document!) your preprocessing here. Your tasks will revolve around words and their connection to sentiment. While preprocessing, keep in mind the domain (restaurant reviews) and the task (sentiment analysis).

### Problem 3. Top words (1 point)
Use a simple word tokenization and count the top 10 words in positive reviews; then the top 10 words in negative reviews*. Once again, try to define what "top" words means. Describe and document your process. Explain your results.

\* Okay, you may want to see top N words (with $N \ge 10$).

### Problem 4. Review titles (2 point)
How do the top words you found in the last problem correlate to the review titles? Do the top 10 words (for each sentiment) appear in the titles at all? Do reviews which contain one or more of the top words have the same words in their titles?

Does the title of a comment present a good summary of its content? That is, are the titles descriptive, or are they simply meant to catch the attention of the reader?

### Problem 5. Bag of words (1 point)
Based on your findings so far, come up with a good set of settings (hyperparameters) for a bag-of-words model for review titles and contents. It's easiest to treat them separately (so, create two models); but you may also think about a unified representation. I find the simplest way of concatenating the title and content too simplistic to be useful, as it doesn't allow you to treat the title differently (e.g., by giving it more weight).

The documentation for `CountVectorizer` is [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html). Familiarize yourself with all settings; try out different combinations and come up with a final model; or rather - two models :).

### Problem 6. Deep sentiment analysis models (1 point)
Find a suitable model for sentiment analysis in English. Without modifying, training, or fine-tuning the model, make it predict all contents (or better, combinations of titles and contents, if you can). Meaure the accuracy of the model compared to the `sentiment` column in the dataset.

### Problem 7. Deep features (embeddings) (1 point)
Use the same model to perform feature extraction on the review contents (or contents + titles) instead of direct predictions. You should already be familiar how to do that from your work on images.

Use the cosine similarity between texts to try to cluster them. Are there "similar" reviews (you'll need to find a way to measure similarity) across different restaurants? Are customers generally in agreement for the same restaurant?

### \* Problem 8. Explore and model at will
In this lab, we focused on preprocessing and feature extraction and we didn't really have a chance to train (or compare) models. The dataset is maybe too small to be conclusive, but feel free to play around with ready-made models, and train your own.